pkg/jobWorker/job.go

```diff
@@ -0,0 +1,307 @@
+
+	// run the command in a Goroutine so that Start can return immediately
+	go func() {
+		// Use cmd.Process.Wait() instead of cmd.Wait() since cmd.Wait() is not thread safe
+		// and we do not want to hold the mutex while waiting for the process to exit.
+		// So instead we use cmd.Process.Wait() and store the result in j.processState to mimic what
+		// cmd.Wait would do.
+		// This prevents concurrency issues when a user calls Start(), the command quickly exits (updating the
+		// process state), and the user invokes Status().
+		processState, err := job.cmd.Process.Wait()
+
+		job.mutex.Lock()
+		defer job.mutex.Unlock()
+
+		job.processState = processState
+		job.status.ExitCode = job.processState.ExitCode()
+		if job.status.State == JOB_STATUS_RUNNING {
+			job.status.State = JOB_STATUS_COMPLETED
+		}
+
+		if err != nil {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error running command: %s\n", err)
+		}
+
+		if err == nil && !job.processState.Success() {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error running command: %s\n", &exec.ExitError{ProcessState: job.processState})
+		}
+
+		// close the output, so that any readers of the output know the process has exited and will no longer
+		// block waiting for new output
+		if err = job.output.Close(); err != nil {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error closing output: %s\n", err)
+		}
+
+		// do not close the cgroup.procs file until after the process has exited
+		if err = ns.DeleteCGroup(cgroupName); err != nil {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error closing cgroup: %s\n", err)
+		}
+
+		if err = ns.UnmountProc(); err != nil {
+			log.Printf("error unmounting /proc - %s\n", err)
+		}
+	}()
+
+	return nil
+}
+
+// Status returns the current Status of the Job.
+func (job *Job) Status() *JobStatus {
+	job.mutex.Lock()
+	defer job.mutex.Unlock()
+	log.Printf("get job status:%s", job)
+	return job.status
+}
+
+// Stream returns an OutputReadCloser that streams the combined stdout and stderr of the Job.
+func (job *Job) Stream() *OutputReadCloser {
+	log.Printf("get job stream:%s", job)
+	return NewOutputReadCloser(job.output)
+}
+
+func (job *Job) Stop() error {
+	job.mutex.Lock()
+	defer job.mutex.Unlock()
+
+	log.Printf("stop job :%s", job)
+	if job.cmd == nil {
+		return ErrJobNotStarted
+	}
+
+	if err := job.cmd.Process.Signal(syscall.SIGTERM); err != nil {
+		return fmt.Errorf("error sending SIGTERM: %w", err)
+	}
+
+	cmdWait := make(chan error, 1)
+	timer := time.NewTimer(10 * time.Second)
+	// stop timer in case process exits before 10 seconds
+	// it's safe to stop timer even if stopped already
+	defer timer.Stop()
+
+	go func() {
+		_, err := job.cmd.Process.Wait()
+		cmdWait <- err
+	}()
+
+	select {
+	case <-cmdWait:
+		// command exited before timer expired, so nothing to do
+	case <-timer.C:
+		// send SIGKILL if process is still running after timer expires
```

Comment:
while waiting, `job.Status()` is blocked.

Reply:
@greedy52 ok. I think I can change Stop method like one the picture below and wait what would happened first SIGTEM or command completion.  

**Explanation**: I don't want to design Stop function as fire-and-forget and wrap all waiting logic in separate go-routine. It definitely would help to prevent blocking all other operations, but Stop method not able to return any error if, by any reason, process would be KILLed .  

<img width="893" alt="Screenshot 2024-09-04 at 11 43 17 AM" src="https://github.com/user-attachments/assets/50451881-dcd0-49c9-9db6-7df59bec885e">

Reply:
What do you think ?

Reply:
I think it depends on the purpose of the lock. If the lock is meant to prevent other API calls, then it's doing its job. If the lock is meant to protect internal members like status, it's not doing anything while waiting so might as well just unlock during that. I will leave it to you to judge that.

Reply:
@greedy52  in general I would agree with you and probably for production solution better to design `Stop` as fire-and-forget and let caller decide - would he like to wait for Termination/Completions or not.

But I'm worrying I would be able to complete challenge on time. If you don't mind - I would add TODO: and if I have time - I would try to review it completing client/server.
 
```go
	err := testJob.Stop()
	for status := testJob.Status(); status.State != JobStatusTerminated && status.State != JobStatusCompleted) {
			fmt.Println(status.State)
			time.Sleep(1 * time.Second)
	}
```
Here is  TODO.

<img width="726" alt="Screenshot 2024-09-05 at 4 04 32 PM" src="https://github.com/user-attachments/assets/8bbfb3ab-e8a1-4ada-b749-403bd133a207">